### DISCUSSION TITLE
Replace specific color with noise

### DISCUSSION THREAD

**Discussion by oarsoy** [ImageMagick/ImageMagick#3866] [Help]
We're trying to replace black color with noise effect. As you see, instead of the black part all photo comes with noise. What's the problem?

Our command:
`convert '/home/public_html/original.png' -fuzz 0% +noise random -opaque black '/home/public_html/result.png'`

![result](https://user-images.githubusercontent.com/43262003/123309736-e8512580-d4ea-11eb-8f7b-1db3ca67db75.jpg)

**Comment by snibgo** [comment]
IM does operations in the order you give them.

"+noise random" changes every pixel to a random value.

"-opaque black" changes all black pixels to the current background colour, which by default is white.

So that is totally different to what you want.

You could do it this way: Clone the image, and "+noise random" that copy. Then composite that noisy image over the original, with a mask that is white where you want the effect and otherwise black. You can make the mask by changing all non-black pixels to white, then "-negate".

**Reply by oarsoy** [reply]
> IM does operations in the order you give them.
> 
> "+noise random" changes every pixel to a random value.
> 
> "-opaque black" changes all black pixels to the current background colour, which by default is white.
> 
> So that is totally different to what you want.
> 
> You could do it this way: Clone the image, and "+noise random" that copy. Then composite that noisy image over the original, with a mask that is white where you want the effect and otherwise black. You can make the mask by changing all non-black pixels to white, then "-negate".

We're not familiar with mask commands. So do we need to convert it with 2 "convert" steps?

**Reply by fmw42** [reply]
He is referring to the use of a binary mask in doing compositing as discussed at https://legacy.imagemagick.org/Usage/compose/#compose

`convert image1 image2 mask -compose over -composite result`

for IM 7 replace convert with magick.

**Comment by snibgo** [comment]
What version of IM, on what platform? I'll assume IM v7, in a Windows BAT script:

Make a test image with black pixels:
```
magick ^
  -size 300x200 radial-gradient: ^
  -level 80,0%% ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/123357889-5422a500-d562-11eb-9af0-f5ed66957616.png)

Make a clone, and make that noisy. Swap those images. For the mask, make another clone of the original, and make non-black pixels white. Do a composite of the noisy image over the original, using the mask.
```
magick ^
  x.png ^
  ( -clone 0 ^
    +noise random ^
  ) ^
  +swap ^
  ( -clone 1 ^
    -fill White +opaque Black ^
  ) ^
  -compose Over -composite ^
  x2.png
```
![x2](https://user-images.githubusercontent.com/33812368/123357897-597fef80-d562-11eb-948a-e41c1cbd92a0.png)
Is that what you want?

EDIT to add: if the input is grayscale (as this test image is) and you want coloured noise, insert "-set colorspace sRGB" after reading x.png.
![x2a](https://user-images.githubusercontent.com/33812368/123358197-f80c5080-d562-11eb-8408-c2e975fe284a.png)

EDIT 2: We can do this without a mask, by making black pixels transparent, and compositing that over the noisy image.
```
magick ^
  x.png ^
  -set colorspace sRGB ^
  -transparent black ^
  ( -clone 0 ^
    -alpha off ^
    +noise random ^
  ) ^
  +swap ^
  -compose Over -composite ^
  x3.png
```
![x3](https://user-images.githubusercontent.com/33812368/123358780-f5f6c180-d563-11eb-9e87-d99aef75cecd.png)

**Reply by oarsoy** [reply]
> ( -clone 0 ^
>     +noise random ^
>   ) ^
>   +swap ^
>   ( -clone 1 ^
>     -fill White +opaque Black ^
>   ) ^
>   -compose Over -composite ^

Thank you for the clear suggestions. Yes, all we want is turns blacks to sRGB colored (preferred) noise like in your example.

We're using this version in our Apache server:
ImageMagick 7.0.11-5 Q16 x86_64 2021-03-28 https://imagemagick.org

So all we need to do is to change the command like that, correct?
`convert '/home/public_html/original.png' -set colorspace sRGB ^ -transparent black ^ ( -clone 0 ^ -alpha off ^ +noise random ^  ) ^ +swap ^ -compose Over -composite ^ '/home/public_html/result.png'`

**Comment by snibgo** [comment]
I suggest "magick". Why do you want to use "convert"?

I suppose your platform is bash. Is that correct? The caret ^ is the Windows line-continuation character. When the caret is the last character on the line, it means the command continues on the next line. The Windows interpreter will strip the caret and the newline from the command before passing it to magick. If you put carets in a bash command, the bash interpreter will simply pass them to magick, which won't know what to do with them.

The bash character for line-continuation is a backslash \.

**Reply by oarsoy** [reply]
> I suggest "magick". Why do you want to use "convert"?
> 
> I suppose your platform is bash. Is that correct? The caret ^ is the Windows line-continuation character. When the caret is the last character on the line, it means the command continues on the next line. The Windows interpreter will strip the caret and the newline from the command before passing it to magick. If you put carets in a bash command, the bash interpreter will simply pass them to magick, which won't know what to do with them.
> 
> The bash character for line-continuation is a backslash .

I'm sorry I'm unfamiliar with Windows platform. We are using ImageMagick on CentOS Linux Apache server with PHP module (imagick module version 3.4.4).

When we execute this command in server or SSH command, it does not work it just creates a file with "^-0" name.

```
exec("magick ^ '/home/public_html/original.png' ^
  -set colorspace sRGB ^
  -transparent black ^
  ( -clone 0 ^
    -alpha off ^
    +noise random ^
  ) ^
  +swap ^
  -compose Over -composite ^ '/home/public_html/result.png'");
```

**Comment by snibgo** [comment]
Bash syntax would be:
```
magick '/home/public_html/original.png' \
  -set colorspace sRGB \
  -transparent black \
  \( -clone 0 \
     -alpha off \
     +noise random \
  \) \
  +swap \
  -compose Over -composite '/home/public_html/result.png'
```
Note that there are no caret characters.
